social_core/backends/microsoft.py

```diff
@@ -45,8 +45,11 @@ def get_user_id(self, details, response):
 
     def get_user_details(self, response):
         """Return user details from Microsoft online account"""
+        email = response.get('mail')
+        if not email and '@' in response.get('userPrincipalName'):
+            email = response.get('userPrincipalName')
         return {'username': response.get('displayName', ''),
-                'email': response.get('mail'),
+                'email': email,
```

Comment:
* Based on the Microsoft documentation email and logon namespaces are basically the same. So, I think you can always use the UPN instead.

```
"The point of the UPN is to consolidate the email and logon namespaces so that the user only needs
to remember a single name."
```
[**User Naming Attributes** - userPrincipalName (Microsoft)](https://msdn.microsoft.com/en-us/library/ms677605(v=vs.85).aspx#userPrincipalName)

Reply:
If I understand correctly, UPN won't always email. In that case, it may cause the wrong entries. From their documentation, it was hard to find the accurate answer as in different app context it was written differently.